### ISSUE TITLE
Mini-batch setting with Semi Markov CRF

### ISSUE THREAD

**Issue by urchade** [harvardnlp/pytorch-struct#110]
I encounter learning instability when using a batch size > 1 with the semi-markovian CRF (loss goes to very large negative number), even when explicitly providing "lengths". I think the bug comes from the masking. 
The model train well when setting batch size 1.

**Comment by srush** [comment]
thanks. Can you by any chance provide an example? I will take a look.

**Comment by urchade** [comment]
Since my work is still an ongoing research project, I can't make it public for now. Can I send it to your email address (srush.research@gmail.com) instead?

**Comment by urchade** [comment]
```python
import torch, torch_struct
import pytorch_lightning as pl
import matplotlib.pyplot as plt

torch.manual_seed(1)

batch, N, C, K = 3, 10, 2, 6

def show_sm(chain):
    plt.imshow(chain.detach().sum(1).sum(-1).transpose(0, 1))

log_potentials = torch.randn(batch, N, K, C, C)

# dist with and withoud mask length (we do not pad the 0th element of the batch)
dist_1 = torch_struct.SemiMarkovCRF(log_potentials)
dist_2 = torch_struct.SemiMarkovCRF(log_potentials, lengths=torch.LongTensor([11, 5, 1]))
dist_3 = torch_struct.SemiMarkovCRF(log_potentials, lengths=torch.LongTensor([11, 5, 4]))

# argmax for the 0th index should be the same for every dist since there is no padding on this index
assert torch.allclose(dist_1.argmax[0], dist_2.argmax[0])
assert torch.allclose(dist_1.argmax[0], dist_3.argmax[0])
assert torch.allclose(dist_2.argmax[0], dist_3.argmax[0])
```

**Comment by srush** [comment]
oh thanks, this is a useful test (and sounds like a bug) 

@da03 we should fix this. Any chance you could take a first look?

**Comment by srush** [comment]
Oh wow, impressive @da03 ! This code is really complex.

Long term let make SemiMarkovParallel and SemiMarkovFlat their own classes and let CRF pick which one to use.
